homework_5/pd_and_visualization.py

```diff
@@ -0,0 +1,420 @@
+of contigs that share the same chromosome number. I will then use query to leave the contigs that are intersected only.
+"""
+
+gff_bed_intersect = gff_for_intersect\
+    .merge(bed_for_intersect, left_on='chromosome', right_on='chromosome') # merged, now there are all combinations
+
+gff_bed_intersect = gff_bed_intersect\
+    .query('start_x > start_y and end_x < end_y') # selected only contigs that intersect with annotation
+
+# let's have a look at the final table
+
+gff_bed_intersect 
+
+# task 5: volcano plot 
+# part 1: preparing
+
+diffexpr = pd.read_csv('diffexpr_data.tsv.gz', sep='\t') # first download the data
+
+# then modify the data: i added two columns on significance and direction of change and then combined them together
+diffexpr['sign_non_sign'] = diffexpr.apply(lambda x: 'Significantly' if x['pval_corr'] < 0.05 else 'Non-significantly', axis=1)
+diffexpr['up_down'] = diffexpr.apply(lambda x: 'downregulated' if x['logFC'] < 0 else 'upregulated', axis=1)
+diffexpr['sector'] = diffexpr['sign_non_sign'] + ' ' + diffexpr['up_down'] # here finally i get the column needed
+diffexpr = diffexpr.sort_values(by=['sign_non_sign', 'up_down'], ascending=[False, True]) # sorted just to please my eyes
+
+# now i have to get some values for plot tuning
+# first one: i need the approximate log_pval value that corresponds to pval_corr=0.05 for a horizontal line
+print(diffexpr[diffexpr['pval_corr'] > 0.05][diffexpr['pval_corr'] < 0.0508]) # finding log_pval for pval_corr = 0.05
+# from this table i choose 1.297575
+
+# secondly, to customize the x axis length i need the maximum and minimum values on the x scale
+print(min(diffexpr['logFC'])) # minimum = -10.661092815248146
+print(max(diffexpr['logFC'])) # maximum = 10.092524279930334
+# okay then i will make the x axis from -11.5 to 11.5
+
+# finally i really need to know top-2 significantly upregulated and significantly downregulated genes
+top_genes = [] # in this list i will put the lists of dots coordinates and gene names
+top_down = diffexpr.query('pval_corr < 0.05').sort_values(by='logFC').iloc[0:2] # top-2 downregulated genes
+top_up = diffexpr.query('pval_corr < 0.05').sort_values(by='logFC', ascending=False).iloc[0:2] # top-2 upregulated genes
+for i in range(2):
+    top_genes.append([top_up.iloc[i].logFC, top_up.iloc[i].log_pval, top_up.iloc[i].Sample])
+    top_genes.append([top_down.iloc[i].logFC, top_down.iloc[i].log_pval, top_down.iloc[i].Sample])
+
+# task 5: volcano plot 
+# part 2: action
+
+plt.rcParams["figure.figsize"] = (8, 5) # figure size
+plt.rcParams['font.weight'] = 'bold'
+plt.rcParams['mathtext.fontset'] = 'custom'
+plt.rcParams['mathtext.bf'] = 'Verdana:italic:bold'
+
+# the plot itself
+res = sns.scatterplot(
+    data=diffexpr, 
+    x='logFC', 
+    y='log_pval', 
+    hue='sector', # for different colors
+    s=4, # dot size
+    linewidth=0 # no white dot lineage
+    ) 
+
+res.axhline(1.297575, c='grey', lw=1, ls='--') # horizontal line
```

Comment:
Хардкодить не надо, а то не понятно, что это
```suggestion
res.axhline(-np.log10(0.05), c='grey', lw=1, ls='--') # horizontal line
```